### PR TITLE
winConsoleUIATextInfo: use rangeFromPoint instead of broken getVisibleRanges

### LINKED ISSUE #10119: Incorrect behavior of "review bottom line" in Windows Console
### Steps to reproduce:
1. Open Command Prompt.
2. Invoke "review bottom line" (desktop layout: shift+numpad 9).

### Actual behavior:
On Windows 10 1803, the last line of output is read as expected, but the review cursor is placed out of bounds: after invoking "review bottom line", blank lines at the end of the output become accessible that previously weren't.

On Windows 10 1903, NVDA says blank. "read current line" also reports blank, but attempting to review the next line says "bottom" followed by the last line of output.

### Expected behavior:
The last line of output is reported and the review cursor is placed there. The review cursor stays within the console bounds (i.e. no blank lines are exposed when the script is invoked).

#### NVDA installed/portable/running from source:
Installed and source.

#### NVDA version:
Latest `master` branch.

#### Windows version:
Windows 10 version 1803 and 1903.

#### Name and version of other software in use when reproducing the issue:
Windows Console (Command Prompt)

#### Does the issue still occur after restarting your PC?
Yes.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No.

### LINKED ISSUE #10191: New UIA support and Braille not working
The support for UIA in console leads to a lot of problems in Braille.

## Steps to reproduce

1. Activate the UIA support for the Windows console in NVDA preferences.
2. Start a console with a Braille display.
3. And browse in review. No need to go far to see the problems.

## Excerpt

Here is an excerpt for my own Windows console.  When I open it, the review cursor is able to read the following:

```
Microsoft Windows [Version 10.0.17763.678]
(c) 2018 Microsoft Corporation. All rights reserved.

C:\Users\Admin>
```

My Braille display, when scrolling with the review cursor (with pavnum7 and pavnum9 on a computer keyboard) reads the following:

```
78]
 Corporation. All rights reserved.

CC:\Users\Admin>
```

(The CC on the last line doesn't appear the first time, it only appears when browsing back to it, which shows some problems.)

So I'd say there is a problem.  And when using the Braille keys to scroll, things get even more weird.  I don't think I need to supply a full transcript: comparing outputs in the previous transcripts should show lots of problems.

So I would highly advise: please don't merge it or do some proper testing in Braille.  Although the feature can be turned off, it's definitely on by default.

I'll be happy to do further testing if need be.

## Information:

- Windows 10 (see signature above if needed)
- NVDA snapshot: `alpha-18590,50619158`
- Braille display: HandyTech ActiveBraille

Thanks!

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #10191 
Fixes #10119

### Summary of the issue:
In Windows consoles with UIA, NVDA currently makes use of IUIAutomationTextPattern::getVisibleRanges to find the bounds of what is visibly on screen. This is both for bounding the review cursor, and for collecting the on-screen text for diffing, in order to speak what has changed.
However, IUIAutomationTextPattern:getVisibleRanges can eventually start giving back junk text ranges that are positioned way above, and outside of the real visible bounds. This seems to start happening once a great deal of text has been scrolled up the screen.
This not only causes the review cursor to act strangely or get stuck when moving to the top or bottom, but it causes very old and invalid text to be spoken when things change on screen.
 
### Description of how this pull request fixes the issue:
Instead of getVisibleRanges, use IUIAutomationTextPattern::rangeFromPoint at the top left and bottom right of the console window.

### Testing performed:
Tested in both wsl (Ubuntu) and cmd.
Tested when older alpha builds of NVDA got into a state where the console no longer made sence with getVisibleRanges; this approach seemed to give back correct text and the review cursor no longer got stuck.

### Known issues with pull request:
Like the older legacy winConsole support, if moving to the bottom with the review cursor and then going up by line, if there are blank lines in the console, these are not skipped. this is a little annoying for efficiency, but it is the same behaviour we originally had for many years, and blank lines are certainly better than  invalid text or getting stuck.

### Change log entry:
None.

### CODE REVIEW

**Comment by michaelDCurran** [comment]
@codeofdusk  I'd appreciate your help in testing this where you can. I was originally against this approach, but since getvisibleRanges is very broken,  I don't think we have a choice anymore.

File: source/NVDAObjects/UIA/winConsoleUIA.py
Comment (leonardder) [inline source/NVDAObjects/UIA/winConsoleUIA.py]: I assume this now works when inherited from the base class, and that's the reason why you removed this?

Comment (michaelDCurran) [reply source/NVDAObjects/UIA/winConsoleUIA.py]: This is correct. The base implementation of _getTextLines seems to be fine as far as I understand.

**Comment by codeofdusk** [comment]
> @codeofdusk I'd appreciate your help in testing this where you can. I was originally against this approach, but since getvisibleRanges is very broken, I don't think we have a choice anymore.

After preliminary testing, I found that:

* The test cases from #9891 and #10406 do not result in broken autorread and a crashed console respectively.
* There is a slight autoread performance hit after the changes (it appears that `_getTextLines` runs a bit slower).
* When a console is flooded with many new text events, it is completely unusable – it is, of course, expected that autoread will stop functioning, but even text review breaks after these changes.

**Comment by michaelDCurran** [comment]
I'm guessing that perhaps when review stops working when the console is 
flooded,  this is when in the existing code new text announced was 
invalid, and review also was broken? Or are you saying that this pr is 
actually worse?

I think I can definitly address the  speed issue by putting back 
_getTextLines that calls IUIAutomationTextRange::getText directly. The 
base _getTextLines breaks text by unit line which is rather slow for 
UIA. Rather as this UIA implementation contains linefeed characters, we 
can just do a python string splitlines.

**Comment by codeofdusk** [comment]
> I'm guessing that perhaps when review stops working when the console is flooded,  this is when in the existing code new text announced was invalid, and review also was broken? Or are you saying that this pr is actually worse? I think I can definitly address the  speed issue by putting back _getTextLines that calls IUIAutomationTextRange::getText directly. The base _getTextLines breaks text by unit line which is rather slow for UIA. Rather as this UIA implementation contains linefeed characters, we can just do a python string splitlines.

This PR is worse. Previously, autoread stopped completely but review still worked.

Also, a `RuntimeError` is thrown when the console exits after this PR.

**Comment by codeofdusk** [comment]
It seems that the `move` method still uses `isTextRangeOffscreen`. Is this intentional?

**Comment by codeofdusk** [comment]
Also, it seems that your `POSITION_LAST` initialization assumes that `_expandCollapseBeforeReview` is `True`. Since #9802 was merged, there might not be a need to do this any more. Consider removing the `_expandCollapseBeforeReview` override.

**Comment by michaelDCurran** [comment]
@codeofdusk  I think I have improved POSITION_FIRST / POSITION_LAST so that now review top and review bottom behave after a lot of text has scrolled up the screen.
I removed the usage of isTextRangeOffscreen, replacing it with calls to compareEndPoints
I have entirely removed _expandCollapseBeforeReview as requested.
I have also fixed the braille issues in #10191 by ensuring that a range is collapsed after a move.
If you could run through your tests again it would be much appreciated.
If there are still major issues after this, I'm afraide we may need to consider not having UIA support for consoles on by default in 2019.3 as there seem to be just too many console bugs we need to work around here.

**Comment by codeofdusk** [comment]
> @codeofdusk I think I have improved POSITION_FIRST / POSITION_LAST so that now review top and review bottom behave after a lot of text has scrolled up the screen.
> I removed the usage of isTextRangeOffscreen, replacing it with calls to compareEndPoints
> I have entirely removed _expandCollapseBeforeReview as requested.
> I have also fixed the braille issues in #10191 by ensuring that a range is collapsed after a move.
> If you could run through your tests again it would be much appreciated.
> If there are still major issues after this, I'm afraide we may need to consider not having UIA support for consoles on by default in 2019.3 as there seem to be just too many console bugs we need to work around here.

Things are a lot better, but:

STR:
1. Open a console.
2. Invoke "review bottom line".
3. Invoke "review current word".

Expected: the last word of the prompt is read out.
Actual: "blank" is reported.

**Comment by codeofdusk** [comment]
> > @codeofdusk I think I have improved POSITION_FIRST / POSITION_LAST so that now review top and review bottom behave after a lot of text has scrolled up the screen.
> > I removed the usage of isTextRangeOffscreen, replacing it with calls to compareEndPoints
> > I have entirely removed _expandCollapseBeforeReview as requested.
> > I have also fixed the braille issues in #10191 by ensuring that a range is collapsed after a move.
> > If you could run through your tests again it would be much appreciated.
> > If there are still major issues after this, I'm afraide we may need to consider not having UIA support for consoles on by default in 2019.3 as there seem to be just too many console bugs we need to work around here.
> 
> Things are a lot better, but:
> 
> STR:
> 
> 1. Open a console.
> 2. Invoke "review bottom line".
> 3. Invoke "review current word".
> 
> Expected: the last word of the prompt is read out.
> Actual: "blank" is reported.

Nevermind, I think that was confusion on my part (as now consoles have blank lines at the end per known issues). The confusing part is that word and line movement have different bounds (i.e. I can access some parts of the console with word review that I can't with line review, and vice versa).

**Comment by michaelDCurran** [comment]
It seems moving by word was a little broken after your pr was merged to 
return a space from the text property. I have worked around this by 
calling IUIAutomationTextRange::getText directly from the word and line 
offset   calculation functions.

**Review by codeofdusk — APPROVED** [review]
Looks good to me!

**Comment by codeofdusk** [comment]
Also, I re-tested the console flooding case. This PR doesn't regress behavior.

**Comment by codeofdusk** [comment]
@michaelDCurran  When do you plan to merge this PR?

**Comment by michaelDCurran** [comment]
Branch policy dictates that I require a review approval from a 
"collaborator" such as @leonardder before I can merge.

File: source/NVDAObjects/UIA/winConsoleUIA.py
Comment (leonardder) [inline source/NVDAObjects/UIA/winConsoleUIA.py]: What happens if we don't do this? We already expanded to line, so the endpoint should be the end of the line already?

File: source/NVDAObjects/UIA/winConsoleUIA.py
Comment (leonardder) [inline source/NVDAObjects/UIA/winConsoleUIA.py]: Consistency with other code

```suggestion
			boundingInfo = self.__class__(obj, textInfos.POSITION_ALL)
```

File: source/NVDAObjects/UIA/winConsoleUIA.py
Comment (leonardder) [inline source/NVDAObjects/UIA/winConsoleUIA.py]: Could you add a small comment that explains what this if statement tries to do? It is hard to read at first glance.

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Just some small things

**Comment by michaelDCurran** [comment]
@leonardder I have improved comments as requested. However, I did not take your suggestion of changing the makeTextInfo call into a direct WinConsoleUIATextInfo construction as this is not needed in the move method -- it can stay high-level. The only reason I did it in WinConsoleUIATextInfo's __init__ method was because I was passing in _rangeObj which makeTextInfo does not allow.

File: source/NVDAObjects/UIA/winConsoleUIA.py
Comment (leonardder) [inline source/NVDAObjects/UIA/winConsoleUIA.py]: Typo:

```suggestion
		# but we are now above or below the visible range,
```

**Review by LeonarddeR — APPROVED** [review]

**Pull request merged**
